### PR TITLE
dm/openapi: use gin as http router instead of echo

### PR DESCRIPTION
**Pull request by Ehco1996**
### What problem does this PR solve? 

as cdc and other projects choose to use gin as http router and https://github.com/deepmap/oapi-codegen/pull/475 was merged(oapi can genereate gin-server code from openapi spec), maybe is good to keep it same.

### What is changed and how it works?

the most change is the code style between two web frameworks,
the biggest different is difficult to implement`error-handling`  in `gin` with middleware way：
`gin. HandlerFunc` not return error 😂 , so i left many `_  = c.Error(err)` in code to make lint pass

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Side effects

 - Increased code complexity

Related changes

### Release note 

```release-note
`None`.
```

### CODE REVIEW

File: dm/dm/master/openapi.go [hunk 1]
```diff
@@ -732,36 +802,37 @@ func (s *Server) DMAPIOperateTableStructure(ctx echo.Context, taskName string, s
...
-	var msg string
-	if tErr, ok := err.(*terror.Error); ok {
-		code = int(tErr.Code())
-		msg = tErr.Error()
-	} else {
-		msg = err.Error()
-	}
-	if sendErr := sendHTTPErrorResp(c, code, msg); sendErr != nil {
-		c.Logger().Error(sendErr)
+func terrorHTTPErrorHandler() gin.HandlerFunc {
```
Comment (Ehco1996) [inline dm/dm/master/openapi.go:817]: I am not confident in the implementation of this middleware... wait for another reviewe with better suggestion

**Comment by asddongmen** [comment]
> /cc @asddongmen

i will review this pr in tomorrow.

File: dm/dm/master/openapi.go (hunk 1 above)
Comment (asddongmen) [reply dm/dm/master/openapi.go:817]: I think it is good.

**Review by asddongmen — APPROVED** [review]

File: dm/dm/master/openapi.go [hunk 2]
```diff
@@ -247,29 +260,39 @@ func (s *Server) DMAPIStartRelay(ctx echo.Context, sourceName string) error {
 	if needUpdate {
 		// update current source relay config before start relay
 		if err := s.scheduler.UpdateSourceCfg(sourceCfg); err != nil {
-			return err
+			_ = c.Error(err)
+			return
 		}
 	}
-	return s.scheduler.StartRelay(sourceName, req.WorkerNameList)
+	if err := s.scheduler.StartRelay(sourceName, req.WorkerNameList); err != nil {
```
Comment (lance6716) [inline dm/dm/master/openapi.go:267]: bad API :joy:  https://github.com/gin-gonic/gin/issues/2941

**Review by lance6716 — APPROVED** [review]

File: Makefile [hunk 3]
```diff
@@ -362,49 +362,49 @@ else
 endif
 
 tools/bin/failpoint-ctl: tools/check/go.mod
-	cd tools/check && $(GO) build -o ../bin/failpoint-ctl github.com/pingcap/failpoint/failpoint-ctl
+	cd tools/check && $(GO) build -mod=mod -o ../bin/failpoint-ctl github.com/pingcap/failpoint/failpoint-ctl
```
[the remark this answers is not in the document]
Comment (Ehco1996) [reply Makefile:365]: `- mod` flag is useful when upgrade pkg version in `tools`

we don't have any go file in tools dir, if i run `go mod tidy` go.mod will be a empty file 😅,do you have any other better way to upgrade those tools ?

**Pull request merged**